spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2012-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.scheduler.Schedulers;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.reactive.LoadBalancerProperties;
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that verifies whether the
+ * instances are alive and only returns the healthy one, unless there are none.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 2.2.0
+ */
+public class HealthCheckServiceInstanceListSupplier
+		implements ServiceInstanceListSupplier {
+
+	private static final Log LOG = LogFactory
+			.getLog(HealthCheckServiceInstanceListSupplier.class);
+
+	private final ServiceInstanceListSupplier delegate;
+
+	private final LoadBalancerProperties loadBalancerProperties;
+
+	private List<ServiceInstance> instances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	private List<ServiceInstance> healthyInstances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	private final InstanceHealthChecker healthChecker;
+
+	public HealthCheckServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			LoadBalancerProperties loadBalancerProperties,
+			InstanceHealthChecker healthChecker) {
+		this.delegate = delegate;
+		this.loadBalancerProperties = loadBalancerProperties;
+		this.healthChecker = healthChecker;
+		initInstances();
+
+	}
+
+	private void initInstances() {
+		delegate.get().subscribe(delegateInstances -> {
+			instances.clear();
+			instances.addAll(delegateInstances);
+		});
+
+		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
+
+		healthCheckFlux.subscribe(verifiedInstances -> {
+			healthyInstances.clear();
+			healthyInstances.addAll(verifiedInstances);
+		});
+	}
+
+	private Flux<List<ServiceInstance>> healthCheckFlux() {
+		return Flux.create(emitter -> Schedulers
+				.newSingle("Health Check Verifier: " + getServiceId(), true)
+				.schedulePeriodically(() -> {
+					List<ServiceInstance> verifiedInstances = new ArrayList<>();
+					Flux.fromIterable(instances).filterWhen(healthChecker::isAlive)
```

Comment:
Change this to a protected method here. Move the implementation from `PingHealthChecker`. If user wants a different implementation, let them extend this class and reuse the rest of the code.

Reply:
ok